aiohomekit/controller/ble/discovery.py

```diff
@@ -65,20 +65,46 @@ def __init__(
         self.controller = controller
         self.device = device
 
-        self.client = BleakClient(self.device)
+        self.client: BleakClient | None = None
+        self._connection_lock = asyncio.Lock()
 
     async def _ensure_connected(self):
-        while not self.client.is_connected:
+        logger.debug("Ensure connected with device %s", self.device)
+        if not self.client:
+            self.client = BleakClient(self.device)
+        if self.client.is_connected:
+            return
+        async with self._connection_lock:
+            while not self.client.is_connected:
+                logger.debug("Connecting to %s", self.device)
+                try:
+                    await self.client.connect()
+                    break
+                except BleakError as e:
+                    logger.debug(
+                        "Failed to connect to %s: %s", self.client.address, str(e)
+                    )
+
+                if self.description.address != self.client.address:
+                    self.client = BleakClient(self.description.address)
+
+                await asyncio.sleep(5)
```

Comment:
This probably shouldn't try forever

Reply:
Yeah it blocks shutdown

2022-07-15 02:54:59.770 ERROR (MainThread) [homeassistant.components.hassio.handler] Timeout on /homeassistant/restart request
2022-07-15 02:55:06.038 DEBUG (MainThread) [aiohomekit.controller.ble.pairing] Failed to connect to FB:EC:42:A9:79:F9: Device with address FB:EC:42:A9:79:F9 was not found.
2022-07-15 02:55:11.040 DEBUG (MainThread) [aiohomekit.controller.ble.pairing] FB:EC:42:A9:79:F9: Connecting
2022-07-15 02:55:21.137 DEBUG (MainThread) [aiohomekit.controller.ble.pairing] Failed to connect to FB:EC:42:A9:79:F9: Device with address FB:EC:42:A9:79:F9 was not found.

Reply:
Opened an issue so we don't forget https://github.com/Jc2k/aiohomekit/issues/98